xchange-bitstamp/src/main/java/org/knowm/xchange/bitstamp/service/BitstampAccountServiceRaw.java

```diff
@@ -14,556 +9,591 @@
-    } else if (currency.equals(Currency.ETH)) {
-      response = withdrawEthFunds(amount, address);
-    } else {
-      throw new ExchangeException(
-          String.format(
-              "Withdrawing funds from Bitstamp failed.Unsupported currency %s", currency));
+    private static final String API_VERSION = "v2";
+
+    private final BitstampDigestV2 signatureCreatorV2;
+
+    private final BitstampAuthenticatedV2 bitstampAuthenticatedV2;
+    private final String apiKeyForV2Requests;
+
+    private final SynchronizedValueFactory<String> uuidNonceFactory;
+    private final SynchronizedValueFactory<String> timestampFactory;
+
+    protected BitstampAccountServiceRaw(Exchange exchange) {
+
+        super(exchange);
+
+        this.bitstampAuthenticatedV2 =
+                ExchangeRestProxyBuilder.forInterface(
+                                BitstampAuthenticatedV2.class, exchange.getExchangeSpecification())
+                        .build();
+
+        this.apiKeyForV2Requests = "BITSTAMP " + exchange.getExchangeSpecification().getApiKey();
+
+        this.signatureCreatorV2 =
+                BitstampDigestV2.createInstance(
+                        exchange.getExchangeSpecification().getSecretKey(),
+                        exchange.getExchangeSpecification().getApiKey());
+
+        BitstampExchange bitstampExchange = (BitstampExchange) exchange;
+
+        this.uuidNonceFactory = bitstampExchange.getUuidNonceFactory();
+        this.timestampFactory = bitstampExchange.getTimestampFactory();
     }
 
-    if (response.error != null) {
-      throw new ExchangeException("Failed to withdraw: " + response.error);
+    public BitstampBalance getBitstampBalance() throws IOException {
+
+        try {
+            BitstampBalance bitstampBalance =
+                    bitstampAuthenticatedV2.getBalance(
+                            apiKeyForV2Requests, signatureCreatorV2, uuidNonceFactory, timestampFactory, API_VERSION);
+            if (bitstampBalance.getError() != null) {
+                throw new ExchangeException("Error getting balance. " + bitstampBalance.getError());
+            }
+
+            return bitstampBalance;
+
+        } catch (BitstampException e) {
+            throw handleError(e);
+        }
     }
 
-    if (response.getId() == null) {
-      return null;
+    public BitstampWithdrawal withdrawBitstampFunds(
+            Currency currency, BigDecimal amount, final String address) throws IOException {
+        return withdrawBitstampFunds(currency, amount, address, null);
     }
 
-    return response;
-  }
+    /**
+     * This method can withdraw any currency if withdrawal endpoint is configured in BitstampAuthenticatedV2
+     **/
+    public BitstampWithdrawal withdrawBitstampFunds(
+            Currency currency, BigDecimal amount, final String address, final String tag)
+            throws IOException {
+        BitstampWithdrawal response;
+
+        amount = amount.setScale(5, RoundingMode.HALF_UP);
```

Comment:
Why was this rounding to 5 decimal places added? Is there some significant figure requirement with the new API? I am not seeing this in the new Bitstamp API documentation?

Reply:
Super, thanks! Tested with BTC too.